File: lib/instagram.ts
import { randomBytes } from "node:crypto";

type InstagramTokenResponse = {
  access_token: string;
  user_id?: number | string;
  token_type?: string;
  permissions?: string[];
};

type InstagramProfile = {
  id: string;
  username?: string;
  account_type?: string;
  media_count?: number;
};

export type InstagramConfig = {
  appId: string;
  appSecret: string;
  redirectUri: string;
  scopes: string;
};

const INSTAGRAM_AUTH_BASE_URL = "https://www.instagram.com/oauth/authorize";
const INSTAGRAM_TOKEN_URL = "https://api.instagram.com/oauth/access_token";
const INSTAGRAM_GRAPH_ME_URL = "https://graph.instagram.com/me";

function requireEnv(name: string): string {
  const value = process.env[name];

  if (!value) {
    throw new Error(`Missing required environment variable: ${name}`);
  }

  return value;
}

export function getInstagramConfig(): InstagramConfig {
  return {
    appId: requireEnv("INSTAGRAM_APP_ID"),
    appSecret: requireEnv("INSTAGRAM_APP_SECRET"),
    redirectUri: requireEnv("INSTAGRAM_REDIRECT_URI"),
    scopes: process.env.INSTAGRAM_SCOPES ?? "instagram_business_basic"
  };
}

export function createInstagramState(): string {
  return randomBytes(24).toString("hex");
}

export function buildInstagramAuthorizeUrl(state: string) {
  const { appId, redirectUri, scopes } = getInstagramConfig();
  const params = new URLSearchParams({
    client_id: appId,
    redirect_uri: redirectUri,
    response_type: "code",
    scope: scopes,
    state
  });

  return `${INSTAGRAM_AUTH_BASE_URL}?${params.toString()}`;
}

export function getInstagramDebugSnapshot(state = "debug-state") {
  const config = getInstagramConfig();

  return {
    appId: config.appId,
    redirectUri: config.redirectUri,
    scopes: config.scopes,
    authorizeUrl: buildInstagramAuthorizeUrl(state)
  };
}

export async function exchangeCodeForToken(code: string) {
  const { appId, appSecret, redirectUri } = getInstagramConfig();
  const body = new URLSearchParams({
    client_id: appId,
    client_secret: appSecret,
    grant_type: "authorization_code",
    redirect_uri: redirectUri,
    code
  });

  const response = await fetch(INSTAGRAM_TOKEN_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded"
    },
    body
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Token exchange failed (${response.status}): ${errorText}`);
  }

  return (await response.json()) as InstagramTokenResponse;
}

export async function fetchInstagramProfile(accessToken: string) {
  const url = new URL(INSTAGRAM_GRAPH_ME_URL);
  url.searchParams.set("fields", "id,username,account_type,media_count");
  url.searchParams.set("access_token", accessToken);

  const response = await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json"
    },
    cache: "no-store"
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Profile fetch failed (${response.status}): ${errorText}`);
  }

  return (await response.json()) as InstagramProfile;
}

File: backend/app/instagram.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from urllib.parse import urlencode

import httpx
from fastapi import HTTPException, status
from itsdangerous import BadSignature, URLSafeSerializer

from .config import Settings


AUTH_URL = "https://www.instagram.com/oauth/authorize"
TOKEN_URL = "https://api.instagram.com/oauth/access_token"
PROFILE_URL = "https://graph.instagram.com/me"
GRAPH_BASE_URL = "https://graph.facebook.com/v23.0"


def build_connect_url(user_id: str, settings: Settings) -> str:
    state = create_state_token(user_id, settings)
    params = {
        "client_id": settings.instagram_app_id,
        "redirect_uri": settings.instagram_redirect_uri,
        "response_type": "code",
        "scope": settings.instagram_scopes,
        "state": state,
    }
    return f"{AUTH_URL}?{urlencode(params)}"


def create_state_token(user_id: str, settings: Settings) -> str:
    serializer = URLSafeSerializer(settings.state_secret, salt="instagram-connect")
    return serializer.dumps({"user_id": user_id})


def parse_state_token(state: str, settings: Settings) -> str:
    serializer = URLSafeSerializer(settings.state_secret, salt="instagram-connect")
    try:
        payload = serializer.loads(state)
    except BadSignature as error:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid Instagram state token",
        ) from error
    return str(payload["user_id"])


async def exchange_code_for_token(code: str, settings: Settings) -> dict[str, object]:
    if not settings.instagram_app_id or not settings.instagram_app_secret:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Instagram OAuth env is not configured",
        )

    body = {
        "client_id": settings.instagram_app_id,
        "client_secret": settings.instagram_app_secret,
        "grant_type": "authorization_code",
        "redirect_uri": settings.instagram_redirect_uri,
        "code": code,
    }

    async with httpx.AsyncClient(timeout=20.0) as client:
        response = await client.post(TOKEN_URL, data=body)

    if response.status_code != 200:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Instagram token exchange failed: {response.text}",
        )

    return response.json()


async def fetch_profile(access_token: str) -> dict[str, object]:
    async with httpx.AsyncClient(timeout=20.0) as client:
        response = await client.get(
            PROFILE_URL,
            params={
                "fields": "id,username,account_type,media_count",
                "access_token": access_token,
            },
        )

    if response.status_code != 200:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Instagram profile fetch failed: {response.text}",
        )
    return response.json()


async def publish_instagram_post(
    access_token: str,
    provider_user_id: str,
    caption: str,
    media_url: str,
    settings: Settings,
) -> dict[str, object]:
    if settings.mock_instagram_publish:
        now = datetime.now(timezone.utc).isoformat()
        return {
            "id": f"mock-media-{provider_user_id}",
            "published_at": now,
            "mode": "mock",
            "caption": caption,
            "media_url": media_url,
        }

    async with httpx.AsyncClient(timeout=25.0) as client:
        create_response = await client.post(
            f"{GRAPH_BASE_URL}/{provider_user_id}/media",
            data={
                "image_url": media_url,
                "caption": caption,
                "access_token": access_token,
            },
        )

        if create_response.status_code != 200:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Instagram media create failed: {create_response.text}",
            )

        creation_id = create_response.json().get("id")
        publish_response = await client.post(
            f"{GRAPH_BASE_URL}/{provider_user_id}/media_publish",
            data={
                "creation_id": creation_id,
                "access_token": access_token,
            },
        )

    if publish_response.status_code != 200:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Instagram media publish failed: {publish_response.text}",
        )

    payload = publish_response.json()
    payload["published_at"] = datetime.now(timezone.utc).isoformat()
    payload["mode"] = "live"
    return payload


def json_dumps(data: object) -> str:
    return json.dumps(data, ensure_ascii=True, sort_keys=True)
